### PR TITLE
[offload] Add support for activation offloading + other changes

### PR DESCRIPTION
**Pull request by anj-s**
# Before submitting

- [x] Was this discussed/approved via a Github issue? (no need for typos, doc improvements)
- [x] Did you read the [contributor guideline](https://github.com/facebookresearch/fairscale/blob/master/CONTRIBUTING.md)?
- [ ] Did you make sure to update the docs?
- [ ] Did you write any new necessary tests?

## What does this PR do?
This PR is for the offload feature. It has the following changes:
* Modify loop to remove the window technique. 
* Add support for activation offloading. This is only partial offloading. I am working on a more complete solution.
* Rename functions and remove comments that contain "rank"
* Add unit tests (just two for now but its a start)

## PR review
Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

File: benchmarks/offload.py [hunk 1]
```diff
@@ -47,31 +48,39 @@ def train(args: argparse.Namespace):
...
-    train_epoch()
+            with torch.autograd.profiler.profile(use_cuda=True, profile_memory=True) as prof:
+                optimizer.zero_grad()
+                inputs = batch_inputs.reshape(-1, args.inputs * args.inputs)
+                with torch.autograd.profiler.record_function("model_training"):
+                    output = model(inputs)
+                    loss = criterion(output, target=batch_outputs)
+                    loss.backward()
+                    optimizer.step()
+            prof.export_chrome_trace("/tmp/mpi_prof")
```
Comment (blefaudeux) [inline benchmarks/offload.py:68]: could that be made a little more generic ? I'm guessing that it's superseded by the other PR actually ?

File: fairscale/nn/misc/offload.py [hunk 2]
```diff
@@ -125,44 +125,74 @@ class ShardSyncLayer(torch.autograd.Function):
...
+            # Move shard from device to offload device.
+            logging.info(f"Dropping shard {drop_index}")
+            model_slices[drop_index].forward_drop()
+
+        if load_index < max_slices:
+            # Load shard from offload device to device.
+            logging.info(f"Loading shard{load_index}")
+            model_slices[load_index].forward_load()
+
+        ctx.inputs = inputs
```
Comment (blefaudeux) [inline fairscale/nn/misc/offload.py:143]: do you really need to save these ?

File: fairscale/nn/misc/offload.py [hunk 3]
```diff
@@ -125,44 +125,74 @@ class ShardSyncLayer(torch.autograd.Function):
...
+        if load_index >= 0:
+            # Load shard from offload device to device.
+            logging.info(f"Backward Loading shard{load_index}")
+            model_slices[load_index].backward_load()
+            model_instance._activations[load_index] = tuple(
+                [a.cuda() for a in list(model_instance._activations[load_index])]
+            )
 
         # The returned variables need to mirror the forward inputs
+        # TODO(anj-s): Why do we need to do this?
```
Comment (blefaudeux) [inline fairscale/nn/misc/offload.py:181]: not sure that it's needed anymore, at some point I was testing with another model which did not handle tuples, feel free to discard

File: benchmarks/offload.py (hunk 1 above)
Comment (anj-s) [reply benchmarks/offload.py:68]: It is. Does the change in the next PR address this?

File: fairscale/nn/misc/offload.py (hunk 3 above)
Comment (anj-s) [reply fairscale/nn/misc/offload.py:181]: I think we still need it and it needs to be equal to the number of inputs. I am going to follow up on this before removing the TODO.

**Pull request merged**